rust/lance/src/dataset/write/merge_insert.rs

```diff
@@ -489,6 +588,262 @@ impl MergeInsertJob {
+                                        *updated_row_addr > *row_addr,
+                                        "Got updated row address that is not in the original batch"
+                                    );
+                                        (0, original_offset)
+                                    }
+                                    _ => (0, original_offset),
+                                }
+                            })
+                            .collect::<Vec<_>>();
+
+                        let updated_batch = interleave_batches(&source_batches, &indices)?;
+
+                        updater.update(updated_batch).await?;
+                    }
+
+                    let updated_fragment = updater.finish().await?;
+                    updated_fragments.lock().unwrap().push(updated_fragment);
+                }
+                Ok(reservation_size)
+            }
+
+            // We shouldn't need much more memory beyond what is already in the batches.
+            let mut memory_size = batches
+                .iter()
+                .map(|batch| batch.get_array_memory_size())
+                .sum();
+
+            loop {
+                let have_additional_cpus = tasks.len() < task_limit;
+                if have_additional_cpus {
+                    if reservation.try_grow(memory_size).is_ok() {
+                        break;
+                    } else if tasks.is_empty() {
+                        // If there are no tasks running, we can bypass the pool limits.
+                        memory_size = 0;
+                        break;
+                    }
+                    // If we can't grow the reservation, we will wait for a task to finish
+                }
+
+                if let Some(res) = tasks.join_next().await {
+                    let size = res??;
+                    reservation.shrink(size);
+                }
+            }
+
+            let fut = handle_fragment(
+                dataset.clone(),
+                fragment,
+                metadata,
+                batches,
+                updated_fragments.clone(),
+                memory_size,
+            );
+            tasks.spawn_on(fut, handle);
+        }
+
+        while let Some(res) = tasks.join_next().await {
+            let size = res??;
+            reservation.shrink(size);
+        }
+        let mut updated_fragments = Arc::try_unwrap(updated_fragments)
+            .unwrap()
+            .into_inner()
+            .unwrap();
+
+        // Collect the updated fragments, and map the field ids. Tombstone old ones
+        // as needed.
+        for fragment in &mut updated_fragments {
+            let updated_fields = fragment.files.last().unwrap().fields.clone();
+            for data_file in &mut fragment.files.iter_mut().rev().skip(1) {
+                for field in &mut data_file.fields {
+                    if updated_fields.contains(field) {
+                        // Tombstone these fields
```

Comment:
Do we tombstone fields in `drop_columns`?

Reply:
We do not right now. We just remove the field from the schema.

The reason I introduced the tombstone is I wanted to be able to move the column without changing the schema. If we change the schema, that means other concurrent update / insert transactions would have to fail with commit conflict, which we don't want. We can't have duplicate ids in the field list of the files, so we need to change the old field locations to -2 so we can use the same field ids in the new files.